File: src/lib/format.ts
import { format as formatDate, formatDistanceToNow } from "date-fns";
import { enUS, zhCN } from "date-fns/locale";
import type { Locale } from "@/shared/locale";
import type { UsageMetric } from "@/shared/usage";

function dateLocale(locale: Locale) {
  return locale === "zh-CN" ? zhCN : enUS;
}

export function formatTime(epoch?: number, locale: Locale = "en"): string {
  if (!epoch) return locale === "zh-CN" ? "暂无" : "No data";
  return formatDate(new Date(epoch * 1000), "MM-dd HH:mm", { locale: dateLocale(locale) });
}

export function relativeTime(epoch?: number, locale: Locale = "en"): string {
  if (!epoch) return locale === "zh-CN" ? "尚未同步" : "Not synced yet";
  return formatDistanceToNow(new Date(epoch * 1000), {
    locale: dateLocale(locale),
    addSuffix: true,
  });
}

function formatNumber(value: number, locale: Locale): string {
  return new Intl.NumberFormat(locale, { maximumFractionDigits: 2 }).format(value);
}

export function hasRemainingSemantics(metric: UsageMetric): boolean {
  return (
    metric.kind === "quota_window" ||
    (metric.limit !== undefined &&
      metric.limit > 0 &&
      (metric.remaining !== undefined || metric.used !== undefined))
  );
}

export function quotaRemaining(metric: UsageMetric): number | undefined {
  if (!hasRemainingSemantics(metric)) return undefined;
  if (metric.remaining !== undefined) return metric.remaining;
  if (metric.limit !== undefined && metric.used !== undefined) {
    return Math.max(0, metric.limit - metric.used);
  }
  if (metric.percentage !== undefined) return Math.max(0, 100 - metric.percentage);
  return undefined;
}

export function quotaRemainingPercentage(metric: UsageMetric): number | undefined {
  if (!hasRemainingSemantics(metric)) return undefined;
  const remaining = quotaRemaining(metric);
  if (remaining !== undefined && metric.limit !== undefined && metric.limit > 0) {
    return Math.min(100, Math.max(0, (remaining / metric.limit) * 100));
  }
  return metric.percentage === undefined
    ? undefined
    : Math.min(100, Math.max(0, 100 - metric.percentage));
}

export function metricValue(metric: UsageMetric, locale: Locale = "en"): string {
  const value = metric.kind === "balance" ? (metric.value ?? metric.used) : metric.used;
  if (value === undefined) return "—";
  if (metric.unit === "%") return `${value.toFixed(1).replace(/\.0$/, "")}%`;
  if (metric.unit === "USD cents") return `$${(value / 100).toFixed(2)}`;
  return `${formatNumber(value, locale)} ${metric.unit}`;
}

export function metricSummary(metric: UsageMetric, locale: Locale = "en"): string {
  if (hasRemainingSemantics(metric)) {
    const remaining = quotaRemaining(metric);
    if (remaining !== undefined) {
      const prefix = locale === "zh-CN" ? "剩余" : "Remaining";
      if (metric.unit === "%") return `${prefix} ${formatNumber(remaining, locale)}%`;
      if (metric.limit !== undefined) {
        return `${prefix} ${formatNumber(remaining, locale)} / ${formatNumber(metric.limit, locale)} ${metric.unit}`;
      }
      return `${prefix} ${formatNumber(remaining, locale)} ${metric.unit}`;
    }
  }
  if (metric.percentage !== undefined)
    return `${metric.percentage.toFixed(1).replace(/\.0$/, "")}%`;
  return metricValue(metric, locale);
}

const englishMetricLabels: Record<string, string> = {
  "codex-credits": "Credits balance",
  "codex-spend-control": "Workspace quota",
  "cursor-api-usage": "API model usage",
  "cursor-auto-usage": "Auto model usage",
  "cursor-grok-bot-weekly": "Grok Bot weekly quota",
  "cursor-models": "Cursor Models",
  "cursor-other-models": "Other Models",
  "cursor-plan-usage": "Included billing-cycle quota",
  "cursor-plan-usage-percent": "Included billing-cycle quota",
  "cursor-usage-based": "On-Demand usage",
  "grok-included-credits": "Included credits quota",
  "grok-included-legacy": "Included credits quota (compatibility)",
  "grok-on-demand": "On-Demand usage",
  "grok-prepaid-balance": "Prepaid credits balance",
  "kimi-booster-balance": "Booster balance",
  "kimi-booster-monthly": "Booster monthly usage",
  "kimi-platform-available-balance": "Available balance",
  "kimi-platform-cash-balance": "Cash balance",
  "kimi-platform-voucher-balance": "Voucher balance",
  "kimi-weekly": "Weekly quota",
  "zenmux-5h": "5-hour Flow quota",
  "zenmux-7d": "7-day Flow quota",
  "zenmux-monthly-cap": "Monthly Flow cap",
};

export function metricLabel(metric: UsageMetric, locale: Locale = "en"): string {
  if (locale === "zh-CN") return metric.label;
  const known = englishMetricLabels[metric.key];
  if (known) {
    if (metric.key === "cursor-auto-usage" && metric.limit !== undefined) return "Auto model quota";
    if (metric.key === "cursor-api-usage" && metric.limit !== undefined) return "API model quota";
    if (metric.key === "cursor-usage-based" && metric.limit !== undefined) return "On-Demand quota";
    if (metric.key === "grok-on-demand" && metric.limit !== undefined) return "On-Demand quota";
    if (metric.key === "kimi-booster-monthly" && metric.limit !== undefined)
      return "Booster monthly quota";
    return known;
  }

  return metric.label
    .replace(/(\d+) 小时窗口/g, "$1-hour window")
    .replace(/(\d+) 天窗口/g, "$1-day window")
    .replace(/(\d+) 分钟额度/g, "$1-minute quota")
    .replace(/(\d+) 小时额度/g, "$1-hour quota")
    .replace(/(\d+) 天额度/g, "$1-day quota")
    .replace(/(\d+) 周额度/g, "$1-week quota")
    .replace(/主窗口/g, "Primary window")
    .replace(/次窗口/g, "Secondary window")
    .replace(/额度窗口 (\d+)/g, "Quota window $1");
}

export function localizedPlan(plan: string | undefined, locale: Locale): string | undefined {
  if (!plan || locale === "zh-CN") return plan;
  const plans: Record<string, string> = {
    "Kimi Code（中国区）": "Kimi Code (China)",
    "Kimi Code（海外区）": "Kimi Code (Global)",
    "Kimi 开放平台（中国区）": "Kimi Platform (China)",
    "Kimi API Platform（海外区）": "Kimi API Platform (Global)",
  };
  return plans[plan] ?? plan;
}

File: client/macos/Sources/UsageCatMenuCore/Format.swift
import Foundation

/// Presentation strings for usage metrics. The client is English-only, so every
/// formatter is pinned to `en-US` rather than the user's locale.
public enum UsageFormat {
    private static let locale = Locale(identifier: "en-US")

    private static let englishMetricLabels: [String: String] = [
        "codex-credits": "Credits balance",
        "codex-spend-control": "Workspace quota",
        "cursor-api-usage": "API model usage",
        "cursor-auto-usage": "Auto model usage",
        "cursor-grok-bot-weekly": "Grok Bot weekly quota",
        "cursor-models": "Cursor Models",
        "cursor-other-models": "Other Models",
        "cursor-plan-usage": "Included billing-cycle quota",
        "cursor-plan-usage-percent": "Included billing-cycle quota",
        "cursor-usage-based": "On-Demand usage",
        "grok-included-credits": "Included credits quota",
        "grok-included-legacy": "Included credits quota (compatibility)",
        "grok-on-demand": "On-Demand usage",
        "grok-prepaid-balance": "Prepaid credits balance",
        "kimi-booster-balance": "Booster balance",
        "kimi-booster-monthly": "Booster monthly usage",
        "kimi-platform-available-balance": "Available balance",
        "kimi-platform-cash-balance": "Cash balance",
        "kimi-platform-voucher-balance": "Voucher balance",
        "kimi-weekly": "Weekly quota",
        "zenmux-5h": "5-hour Flow quota",
        "zenmux-7d": "7-day Flow quota",
        "zenmux-monthly-cap": "Monthly Flow cap",
    ]

    /// Labels that read as a quota rather than raw consumption once a limit exists.
    private static let quotaLabelOverrides: [String: String] = [
        "cursor-auto-usage": "Auto model quota",
        "cursor-api-usage": "API model quota",
        "cursor-usage-based": "On-Demand quota",
        "grok-on-demand": "On-Demand quota",
        "kimi-booster-monthly": "Booster monthly quota",
    ]

    private static let planTranslations: [String: String] = [
        "Kimi Code（中国区）": "Kimi Code (China)",
        "Kimi Code（海外区）": "Kimi Code (Global)",
        "Kimi 开放平台（中国区）": "Kimi Platform (China)",
        "Kimi API Platform（海外区）": "Kimi API Platform (Global)",
    ]

    private static let providerTitles: [String: String] = [
        "codex": "Codex",
        "cursor": "Cursor",
        "grok": "Grok",
        "kimi": "Kimi",
        "zenmux": "ZenMux",
    ]

    // MARK: - Quota math

    public static func hasRemainingSemantics(_ metric: UsageMetric) -> Bool {
        if metric.metricKind == .quotaWindow { return true }
        if let limit = metric.limit, limit > 0, metric.remaining != nil || metric.used != nil {
            return true
        }
        return false
    }

    public static func quotaRemaining(_ metric: UsageMetric) -> Double? {
        guard hasRemainingSemantics(metric) else { return nil }
        if let remaining = metric.remaining { return remaining }
        if let limit = metric.limit, let used = metric.used {
            return max(0, limit - used)
        }
        if let percentage = metric.percentage {
            return max(0, 100 - percentage)
        }
        return nil
    }

    public static func quotaRemainingPercentage(_ metric: UsageMetric) -> Double? {
        guard hasRemainingSemantics(metric) else { return nil }
        if let remaining = quotaRemaining(metric), let limit = metric.limit, limit > 0 {
            return min(100, max(0, (remaining / limit) * 100))
        }
        if let percentage = metric.percentage {
            return min(100, max(0, 100 - percentage))
        }
        return nil
    }

    /// Bars show headroom for quotas and raw progress for everything else.
    public static func progressValue(_ metric: UsageMetric) -> Double? {
        hasRemainingSemantics(metric) ? quotaRemainingPercentage(metric) : metric.percentage
    }

    public static func progressFraction(_ metric: UsageMetric) -> Double? {
        progressValue(metric).map { min(max($0 / 100, 0), 1) }
    }

    // MARK: - Labels

    public static func metricLabel(_ metric: UsageMetric) -> String {
        guard let known = englishMetricLabels[metric.key] else { return metric.label }
        if metric.limit != nil, let quotaLabel = quotaLabelOverrides[metric.key] {
            return quotaLabel
        }
        return known
    }

    public static func localizedPlan(_ plan: String?) -> String? {
        guard let plan else { return nil }
        return planTranslations[plan] ?? plan
    }

    public static func providerTitle(_ provider: String) -> String {
        providerTitles[provider] ?? provider
    }

    // MARK: - Values

    public static func metricValue(_ metric: UsageMetric) -> String {
        let amount = metric.metricKind == .balance ? (metric.value ?? metric.used) : metric.used
        guard let amount else { return "—" }
        return qualifiedAmount(amount, unit: metric.metricUnit)
    }

    public static func metricSummary(_ metric: UsageMetric) -> String {
        if hasRemainingSemantics(metric), let remaining = quotaRemaining(metric) {
            if metric.metricUnit == .percent {
                return "Remaining \(formatNumber(remaining))%"
            }
            if let limit = metric.limit {
                return "Remaining \(formatNumber(remaining)) / \(formatNumber(limit)) \(metric.unit)"
            }
            return "Remaining \(formatNumber(remaining)) \(metric.unit)"
        }
        if let percentage = metric.percentage {
            return "\(compactPercent(percentage))%"
        }
        return metricValue(metric)
    }

    public static func leftoverPhrase(_ metric: UsageMetric) -> String {
        if let percent = quotaRemainingPercentage(metric) {
            return "\(compactPercent(percent))% left"
        }
        if let remaining = quotaRemaining(metric) {
            return "\(bareAmount(remaining, unit: metric.metricUnit)) left"
        }
        if let value = metric.value ?? metric.used {
            return "\(bareAmount(value, unit: metric.metricUnit)) left"
        }
        return metricSummary(metric)
    }

    public static func resetPhrase(_ epoch: Int?) -> String? {
        guard let epoch else { return nil }
        let date = Date(timeIntervalSince1970: TimeInterval(epoch))
        let calendar = Calendar.current
        let time = formatted(date, format: "H:mm")
        if calendar.isDateInToday(date) {
            return "Resets \(time)"
        }
        let sameYear = calendar.component(.year, from: date) == calendar.component(.year, from: Date())
        return "Resets \(formatted(date, format: sameYear ? "MMMM d" : "MMMM d, yyyy")) at \(time)"
    }

    // MARK: - Primitives

    static func formatNumber(_ value: Double) -> String {
        let formatter = NumberFormatter()
        formatter.locale = locale
        formatter.maximumFractionDigits = 2
        formatter.minimumFractionDigits = 0
        return formatter.string(from: NSNumber(value: value)) ?? String(value)
    }

    static func compactPercent(_ value: Double) -> String {
        let text = String(format: "%.1f", value)
        return text.hasSuffix(".0") ? String(text.dropLast(2)) : text
    }

    /// Spells the unit out, for standalone values.
    private static func qualifiedAmount(_ value: Double, unit: MetricUnit) -> String {
        switch unit {
        case .percent:
            return "\(compactPercent(value))%"
        case .usdCents:
            return String(format: "$%.2f", value / 100)
        case .other(let name):
            return "\(formatNumber(value)) \(name)"
        }
    }

    /// Drops the unit name, for captions that already sit under a labelled bar.
    private static func bareAmount(_ value: Double, unit: MetricUnit) -> String {
        switch unit {
        case .percent:
            return "\(formatNumber(value))%"
        case .usdCents:
            return String(format: "$%.2f", value / 100)
        case .other:
            return formatNumber(value)
        }
    }

    private static func formatted(_ date: Date, format: String) -> String {
        let formatter = DateFormatter()
        formatter.locale = locale
        formatter.dateFormat = format
        return formatter.string(from: date)
    }
}
